### PR TITLE
Rotation State Management

### PR DESCRIPTION
**Pull request by geniusgeek**
The project have been modified to Handle Rotation Issues For recyclerview-multiselect

### CODE REVIEW

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java [hunk 1]
```diff
@@ -49,7 +49,8 @@
      */
     public SwappingHolder(View itemView, MultiSelector multiSelector) {
         super(itemView, multiSelector);
-
+        this.itemView.setOnLongClickListener(this);
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: This is a really useful time saver in individual projects, but it wouldn't be a good idea to force it on others who might be using this code. They might want to start up a multiselection mode in other ways besides a long press on the top level itemView.

How about an additional class that adds this extra bit of wiring in? That would make it easier as an out of the box solution, but still make it possible to accomplish stranger goals.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java [hunk 2]
```diff
@@ -284,4 +285,29 @@ private static StateListAnimator getRaiseStateListAnimator(Context context) {
...
+     *  just few cleanup to handle starting up the selection mode
+     *  Note: subclass should   call super.onLongClick before implementation of their onLongClick
+     */
+    @Override
+    public boolean onLongClick(View v) {
+        beginSelectionActionMode();
+        mMultiSelector.setSelected(this, true);
+        return true;
+    }
+    public abstract void beginSelectionActionMode();
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: This would need a good Javadoc, too.

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java [hunk 3]
```diff
@@ -45,12 +46,32 @@ public void onCreate(Bundle savedInstanceState) {
...
+        }
+    }
+
     @TargetApi(11)
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup parent, Bundle savedInstanceState) {
         View v = inflater.inflate(R.layout.fragment_recyclerview, parent, false);
 
-        if (mSubtitleVisible) {
+         if (mSubtitleVisible) {
```
Comment (jingibus) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: These little spacing changes should be reverted. Spacing must be consistent throughout the project.

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java [hunk 4]
```diff
@@ -45,12 +46,32 @@ public void onCreate(Bundle savedInstanceState) {
...
+     *  However after having second thoughts, since the fragment is retained then all the states and instance variables are
+     *  retained as well. no need to make the selection states percelable therefore just check for the selectionstate
+     *  from the multiselector
+     */
+    @Override
+    public void onActivityCreated(Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+        if(mMultiSelector.isSelectable()){
+            if(mDeleteMode !=null){
+                mDeleteMode.setClearOnPrepare(!mDeleteMode.shouldClearOnPrepare());
```
Comment (jingibus) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: Why is this being inverted here?

**Comment by jingibus** [comment]
Thank you for the PR.

Check out my comments; you're doing a few different things here, each of which probably merits its own PR. 

The big thing is the rotation handling in the sample code. Updating that code to rotate correctly with setRetainInstance(true) is great, but it's not going to be enough for most people, who probably won't be using a retained fragment.

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java (hunk 4 above)
Comment (geniusgeek) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: just to avoid clearing the selectable state before starting the action mode after rotating when in selection mode. 
i did this because the clearing on selection mode is set to true by default

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java (hunk 3 above)
Comment (geniusgeek) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: got it.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java (hunk 2 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: got it.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java (hunk 1 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: Do you mean a SelectionManager interface that does functions like?:

 public abstract void startSelection();
    public abstract void stopSelection();
    public void unselectAll();
    public void selectAll();
    public abstract void beginSelectionActionMode();

Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: please reply

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.bignerdranch.android.multiselector;
+
+import java.io.Serializable;
+import java.lang.ref.WeakReference;
+
+/**
+ * Created by Genius on 1/30/2015.
+ */
+public class WeakReferenceSelectableHolder<T extends  SelectableHolder> extends WeakReference implements Serializable{
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java]: This class cannot safely implement Serializable. It will compile correctly, but if this were to actually be serialized, it would blow up, because neither WeakReference nor SelectableHolder are safe to serialize.

Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java]: Luckily, that's okay — as mentioned above, this does not actually need to be serialized, so this class can safely go away.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakHolderTracker.java [hunk 6]
```diff
@@ -1,14 +1,16 @@
...
 import android.util.SparseArray;
 
 import java.lang.ref.WeakReference;
 import java.util.ArrayList;
 import java.util.List;
 
-class WeakHolderTracker {
-    private SparseArray<WeakReference<SelectableHolder>> mHoldersByPosition =
-            new SparseArray<WeakReference<SelectableHolder>>();
+class WeakHolderTracker implements Parcelable {
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakHolderTracker.java]: This class's role is to keep track of currently active ViewHolders. ViewHolders can't safely be serialized or parcelled, so this class should not implement Parcelable either.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java (hunk 5 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java]: so what do you recommend we do concerning the rotation issue if the fragment is not retained?

Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java]: It's only this class that is not safe to serialize. Not everything in MultiSelector must be serialized/parcelled to implement those interfaces.

A better strategy, though, would probably be to create a separate class to hold the serializable parts of MultiSelector:
    
    SelectionState state = mMultiSelector.getSelectionState();
    savedInstanceState.putParcelable(KEY_SELECTIONS, state);

    ...

    SelectionState state = (SelectionState) savedInstanceState.getParcelable(KEY_SELECTIONS);
    mMultiSelector.setSelectionState(state);

It'd be a little manual, but once that piece is built, other useful things could be built with it.

Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java]: ok . i will give it a try . Also im sorry for the stress ive give you due to the whitespaces. im aware of that now. and will fix it. just bear with me. im new to git.

Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/WeakReferenceSelectableHolder.java]: No problem. Let me know if I'm being unclear.

**Comment by geniusgeek** [comment]
yes !!! i have done it better. please take a look at the code that maintains the state of the multiselector. It has been documented At least

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java [hunk 7]
```diff
@@ -27,24 +28,88 @@
...
     private boolean mSubtitleVisible;
-    
+
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setHasOptionsMenu(true);
         getActivity().setTitle(R.string.crimes_title);
-        setRetainInstance(true);
+        //setRetainInstance(true);
```
Comment (jingibus) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: This can just be deleted rather than commented out.

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java [hunk 8]
```diff
@@ -27,24 +28,88 @@
...
+     * However after having second thoughts, since the fragment is retained then all the states and instance variables are
+     * retained as well. no need to make the selection states percelable therefore just check for the selectionstate
+     * from the multiselector
+     */
+    @Override
+    public void onActivityCreated(Bundle savedInstanceState) {
+
+        if (mMultiSelector != null) {
+            Bundle bundle = savedInstanceState;
+            if (bundle != null)
```
Comment (jingibus) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: Include curly braces for this if statement.

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java [hunk 9]
```diff
@@ -192,8 +222,10 @@ public CrimeHolder(View itemView) {
             mDateTextView = (TextView) itemView.findViewById(R.id.crime_list_item_dateTextView);
             mSolvedCheckBox = (CheckBox) itemView.findViewById(R.id.crime_list_item_solvedCheckBox);
             itemView.setOnClickListener(this);
+            itemView.setClickable(true);
             itemView.setOnLongClickListener(this);
-            itemView.setLongClickable(true);
```
Comment (jingibus) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: What did these two changes accomplish? Does long-clicking items still activate selection? Was tapping items turning selection on and off?

File: recyclerview-multiselect/src/main/java/android/support/v7/widget/RebindReportingHolder.java [hunk 10]
```diff
@@ -46,8 +46,13 @@ private void checkFlags(int setFlags) {
...
     }
 
+    /**
+     * check if the view is due for rebiding
+     * @param flag
+     * @return
+     */
     private static boolean isRelevantFlagSet(int flag) {
-        for (Integer value : new int[] { FLAG_BOUND, FLAG_CHANGED, FLAG_UPDATE, FLAG_RETURNED_FROM_SCRAP }) {
+        for (Integer value : new int[] { FLAG_BOUND, FLAG_CHANGED, FLAG_UPDATE, FLAG_RETURNED_FROM_SCRAP ,FLAG_TMP_DETACHED}) {
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/android/support/v7/widget/RebindReportingHolder.java]: If this fixes a real issue, it should be in a separate PR. Does not seem to be related to the selection work you have done here.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java [hunk 11]
```diff
@@ -32,11 +33,25 @@
...
-public class MultiSelector {
+public class MultiSelector extends SelectionManager {
+    public static final String TAG = "multiselector";
+    private static final String SELECTION_POSITIONS = "position";
+    private static final String SELECTIONS_STATE = "state";
     private SparseBooleanArray mSelections = new SparseBooleanArray();
     private WeakHolderTracker mTracker = new WeakHolderTracker();
-
     private boolean mIsSelectable;
+    private boolean mShdSelectAll;
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: "Select all" functionality should be in a separate PR. This PR is for saving/loading selection state.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java [hunk 12]
```diff
@@ -32,11 +33,25 @@
  * }
  * </pre>
  */
-public class MultiSelector {
+public class MultiSelector extends SelectionManager {
+    public static final String TAG = "multiselector";
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: This constant should not be here.

1. If it is used for LogCat, well — library code should not be logging to LogCat. It's fine to do for debugging, but we do not want to clutter up our users' device logs without a really good reason.
2. This is also used for savedInstanceState; those tags should be defined in the Fragment/Activity/etc, not here.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java [hunk 13]
```diff
@@ -196,4 +205,64 @@ private void refreshHolder(SelectableHolder holder) {
...
+    /**
+     * @return Bundle containing the states of the selection and a flag indicating if the multiselection is in
+     * selection mode or not
+     */
+    @Override
+    public Bundle saveSelectionStates() {
+        Bundle information = new Bundle();
+        information.putIntegerArrayList(SELECTION_POSITIONS, (ArrayList<Integer>) getSelectedPositions());
+        information.putBoolean(SELECTIONS_STATE, isSelectable());
+        return information;
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: :+1: Looks great!

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java [hunk 14]
```diff
@@ -196,4 +205,64 @@ private void refreshHolder(SelectableHolder holder) {
...
+    }
+
+    private void restoreSelections(List<Integer> selected) {
+        if (selected == null) return;
+        int position;
+        for (int i = 0; i < selected.size(); i++) {
+            position = selected.get(i);
+            mSelections.put(position, true);
+            mTracker.bindHolder(mTracker.getHolder(position), position);
+        }
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: 1. If mSelections has stale data, then it might not have exactly what was passed in at the beginning of this method. So mSelections.clear() should be called before entering the for() loop here.

2. Instead of calling mTracker.bindHolder() here inside the for loop, call refreshAllHolders() after the for loop has run.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectableHolder.java [hunk 15]
```diff
@@ -3,13 +3,7 @@
 /**
  * <p>Public interface used by selectable items connected to {@link com.bignerdranch.android.multiselector.MultiSelector}.</p>
  */
-public interface SelectableHolder {
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectableHolder.java]: I don't mind leaving these changes here. In the future, though, I strongly recommend looking critically at the code you commit to make sure you are making meaningful changes. This file has changes in spacing and method ordering, but nothing material has changed. 

It pays to be intentional about these things: maybe it is a good idea to change the spacing and method ordering. If so, do that intentionally in a separate commit, in a separate PR. The folks you work with will appreciate it. 

(And I appreciate your work, too! By the way.)

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectionManager.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.support.v7.app.ActionBarActivity;
+import android.view.ActionMode;
+
+/**
+ * Created by Genius on 2/2/15.
+ */
+public abstract class SelectionManager {
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectionManager.java]: This PR is about fixing rotation issues, and this class does not seem to have anything to do that. So I recommend deleting this class and associated changes from the PR. Recommend creating a separate PR for this work if you think it's worthwhile.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java [hunk 17]
```diff
@@ -16,35 +16,42 @@
...
  * <p>selectionModeBackgroundDrawable defaults to a StateListDrawable that displays
  * your colorAccent theme color when state_activated=true, and nothing otherwise.
  * selectionModeStateListAnimator defaults to a raise animation that animates selection
  * items to a 12dp translationZ.</p>
- *
+ * <p/>
  * <p>(Thanks to <a href="https://github.com/kurtisnelson/">Kurt Nelson</a> for examples and discussion on approaches here.)</p>
  */
-public class SwappingHolder extends MultiSelectorBindingHolder implements SelectableHolder {
+public abstract class SwappingHolder extends MultiSelectorBindingHolder implements SelectableHolder {
```
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: This class should not be abstract.

Apart from that change, I am fine with merging this as-is. 

I'll make a similar remark here as I made above about SelectableHolder: there are a lot of changes to this class that I just read through, but after taking the time to understand them it looks like no code actually changed. Kind of confusing for me when this PR is supposed to be about saving out selection state.

**Comment by jingibus** [comment]
:+1: 

Looking good! Very close to merging this.

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java (hunk 9 above)
Comment (geniusgeek) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: long click of items still activates selection. don't worry. i was just activating click events for the the itemView. When user clicks the view. the states is set to selected.

Comment (geniusgeek) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: i felt that since we have click events set for the view. we just need to set the view to be clickable

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java (hunk 12 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: got it

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java (hunk 11 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: ok.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java (hunk 14 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: noted. However even if the mSelections had stale data. it will be replaced by the current data, so i don't think this affects the performance in any way. however. i have cleared it before the loop

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectionManager.java (hunk 16 above)
Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectionManager.java]: The selector manager is the abstract class that provides the interface for selecting with action mode, saving states, restoring states, etc . Select all is just an additional feature, Users can extend this to add their unique features. However i feel that we should remove only the selectAll method instead of the whole class. What do you think?

File: criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java (hunk 9 above)
Comment (jingibus) [inline criminalintent-sample/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.java]: Does it fix anything, though? Did it not work prior to these changes?

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectionManager.java (hunk 16 above)
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SelectionManager.java]: Well, remember: this PR is for rotation state management, not for adding a select all feature or any of that. SelectionManager doesn't need to exist for MultiSelector to have those methods, either. 

I'm not ruling out including that functionality in the project ever; but its presence here is unrelated to state management, and is keeping us from marking that "done" and moving on to other things.

File: recyclerview-multiselect/build.gradle [hunk 18]
```diff
@@ -35,48 +35,6 @@ android {
...
-                    developer {
-                        id 'jingibus'
-                        name 'Bill Phillips'
-                        email 'phillips@bignerdranch.com'
-                    }
-                }
-            }
-        }
-    }
-}
```
Comment (jingibus) [inline recyclerview-multiselect/build.gradle]: This section needs to be restored before merge.

To get the project to compile with this section intact, add the following lines to a gradle.properties file at the top level of the project:

    ossrhUsername=unused
    ossrhPassword=unused

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java (hunk 11 above)
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: This property still needs to be removed before merge.

Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: i have done it. its working fine.
I think its worth mentioning that my run config had issues: it flagged
warning recognizing criminal intent as a module instead of an android
project. what do you think i should do?

Comment (geniusgeek) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/MultiSelector.java]: THanks . I've reverted the file and cleaned up the project. take a look.
sorry for the mistake.

File: recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java (hunk 17 above)
Comment (jingibus) [inline recyclerview-multiselect/src/main/java/com/bignerdranch/android/multiselector/SwappingHolder.java]: This class still needs to be marked non-abstract.

**Comment by jingibus** [comment]
Rad! Make that one change, and it's ready to merge.

**Comment by geniusgeek** [comment]
Making this class a concrete type will be violating one of "Design
Principles" - Favor Composition over inheritance. The class just acts as an
interface/API . However i will do it if you want.

**Comment by jingibus** [comment]
Making it abstract also violates that design principle, because the class must be inherited to be used. Any part of the API that subclasses ViewHolder would do the same thing.

Definitely open to ideas for the API that can allow folks to use composition instead of inheritance, though.

**Comment by geniusgeek** [comment]
thanks

**Comment by geniusgeek** [comment]
please i have a question: ive been thinking a lot of gabage collection in
java and android. Please i need to know the instances in android when
gabage collection can occur. you can use the recyclerviewmultiselector as
example.

**Comment by geniusgeek** [comment]
hi bill i was thinking if we could add a raise animator for prelolipop
devices into the selector library, so i created a shape drawable with
gradient 80333333 that can be applied to a view directly below the selected
view  but not sure how to dynamically add this

**Comment by jingibus** [comment]
The trick is that there's no automatic mechanism for performing the animation, since there are no StateListAnimators. So the animations would need to be triggered and handled inside of setActivated.

**Comment by geniusgeek** [comment]
Hi Bill please take a look at the commit i made concerning the
stateListAnimator. I think its working fine
